KOGITO-3828 - new monitoring addon automatic wiring of listeners

r00ta opened the pull request:
Jira: https://issues.redhat.com/browse/KOGITO-3828

This PR aims to refactor the wiring of the drools/process listener for the monitoring. 

The new module structure looks like the following: 

![Screenshot from 2020-11-25 16-34-34](https://user-images.githubusercontent.com/18282531/100248747-229a1a80-2f3c-11eb-9869-6c1060b81961.png)

If the user imports the `monitoring-core-quarkus` or the `monitoring-core-springboot` addon, then the default listeners for rules and processes will be created unless the user specifies the following config property in his `application.properties` file: 
```
kogito.addon.monitoring.process.deafult.bean.create=false
kogito.addon.monitoring.rule.deafult.bean.create=false
```

The same behaviour is guaranteed if the user imports the `monitoring-prometheus-quarkus` or the `monitoring-prometheus-springboot` module. 

Many thanks for submitting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](CONTRIBUTING.md)
- [x] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [x] Pull Request title contains the target branch if not targeting master: `[0.9.x] KOGITO-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

r00ta commented:
Jenkins retest this please

r00ta commented:
See https://github.com/kiegroup/kogito-examples/pull/458 for the updated examples

cristianonicolai commented:
jenkins please retest this

r00ta commented:
Hi @danielezonca @MarianMacik wdyt about this one?

danielezonca commented on addons/monitoring-addon/monitoring-core/monitoring-core-quarkus/src/main/java/org/kie/kogito/monitoring/core/quarkus/QuarkusEventListenerFactory.java:
```diff
@@ -0,0 +1,32 @@
+    private static final Logger LOGGER = LoggerFactory.getLogger(QuarkusEventListenerFactory.class);
+
+    @Produces
+    @IfBuildProperty(name = "kogito.addon.monitoring.rule.deafult.bean.create", stringValue = "true", enableIfMissing = true)
```
Can you please use a constant? Please use a shorter property name, I think something like `kogito.monitoring.rule.useDefault` could be enough. Wdyt?

r00ta replied:
Done

danielezonca commented on addons/monitoring-addon/monitoring-core/monitoring-core-quarkus/src/main/java/org/kie/kogito/monitoring/core/quarkus/QuarkusEventListenerFactory.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    @Produces
+    @IfBuildProperty(name = "kogito.addon.monitoring.process.deafult.bean.create", stringValue = "true", enableIfMissing = true)
```
Same as above

r00ta replied:
Done

danielezonca commented on addons/monitoring-addon/monitoring-core/monitoring-core-springboot/pom.xml:
```diff
@@ -0,0 +1,40 @@
+  <dependencies>
+    <dependency>
+      <groupId>javax.annotation</groupId>
+      <artifactId>javax.annotation-api</artifactId>
```
Where is it used?

r00ta replied:
Removed

danielezonca commented on addons/monitoring-addon/monitoring-prometheus-addon/monitoring-prometheus-common/pom.xml:
```diff
@@ -0,0 +1,45 @@
+
+    <dependency>
+      <groupId>org.jboss.spec.javax.ws.rs</groupId>
+      <artifactId>jboss-jaxrs-api_2.1_spec</artifactId>
```
Please move rest to platform specific modules: we are reviewing rest dep to use spring/quarkus specific (see https://github.com/kiegroup/kogito-runtimes/pull/891 ). You can keep here an abstract if needed

r00ta replied:
Done

danielezonca reviewed (commented):
Please review beans.xml files, I think there are too many of them

r00ta commented on addons/monitoring-addon/monitoring-core/monitoring-core-common/pom.xml:
```diff
@@ -0,0 +1,79 @@
+      <artifactId>cdi-api</artifactId>
+      <version>2.0.SP1</version>
+      <scope>provided</scope>
+    </dependency>
```
Removed

r00ta commented:
Hi @danielezonca , thanks for the tips, I've changed the files accordingly. Let me know what do you think. 

Regarding the beans.xml files I removed a couple of them that were not needed anymore.

danielezonca commented on addons/monitoring-addon/monitoring-core/monitoring-core-common/pom.xml:
```diff
@@ -0,0 +1,72 @@
+  </parent>
+  <modelVersion>4.0.0</modelVersion>
+  <name>Kogito :: Add-Ons :: Monitoring Core Api</name>
+  <dependencyManagement>
+    <dependencies>
+      <dependency>
+        <groupId>io.quarkus</groupId>
+        <artifactId>quarkus-bom</artifactId>
+        <version>${version.io.quarkus}</version>
+        <type>pom</type>
+        <scope>import</scope>
+      </dependency>
+    </dependencies>
```
This should not be added here that is a common module but in the platform specific

r00ta replied:
Thanks for the catch!

danielezonca commented on addons/monitoring-addon/monitoring-core/monitoring-core-springboot/src/main/java/org/kie/kogito/monitoring/core/springboot/SpringbootProcessEventListenerFactory.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Bean
+    public DefaultProcessEventListenerConfig produceProcessListener() {
+        LOGGER.info("Producing default listener for process monitoring.");
```
```suggestion
        LOGGER.debug("Producing default listener for process monitoring.");
```

danielezonca commented on addons/monitoring-addon/monitoring-core/monitoring-core-springboot/src/main/java/org/kie/kogito/monitoring/core/springboot/SpringbootRuleEventListenerFactory.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Bean
+    public DefaultRuleEventListenerConfig produceRuleListener() {
+        LOGGER.info("Producing default listener for rule monitoring.");
```
```suggestion
        LOGGER.debug("Producing default listener for rule monitoring.");
```

danielezonca commented on addons/monitoring-addon/monitoring-core/monitoring-core-quarkus/src/main/java/org/kie/kogito/monitoring/core/quarkus/QuarkusEventListenerFactory.java:
```diff
@@ -0,0 +1,48 @@
+    @Produces
+    @IfBuildProperty(name = Constants.MONITORING_RULE_USE_DEFAULT, stringValue = "true", enableIfMissing = true)
+    public DefaultRuleEventListenerConfig produceRuleListener() {
+        LOGGER.info("Producing default listener for rule monitoring.");
```
```suggestion
        LOGGER.debug("Producing default listener for rule monitoring.");
```

danielezonca commented on addons/monitoring-addon/monitoring-core/monitoring-core-quarkus/src/main/java/org/kie/kogito/monitoring/core/quarkus/QuarkusEventListenerFactory.java:
```diff
@@ -0,0 +1,48 @@
+    @Produces
+    @IfBuildProperty(name = Constants.MONITORING_PROCESS_USE_DEFAULT, stringValue = "true", enableIfMissing = true)
+    public DefaultProcessEventListenerConfig produceProcessListener() {
+        LOGGER.info("Producing default listener for process monitoring.");
```
```suggestion
        LOGGER.debug("Producing default listener for process monitoring.");
```

danielezonca commented on addons/monitoring-addon/monitoring-core/pom.xml~HEAD:
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This file is probably the output of a merge conflict addons/monitoring-addon/monitoring-core/pom.xml**~HEAD**

r00ta replied:
wooops! Thanks for the catch!

danielezonca commented on addons/monitoring-addon/monitoring-core/pom.xml~upstream_master:
```diff
@@ -0,0 +1,63 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This file is probably the output of a merge conflict addons/monitoring-addon/monitoring-core/pom.xml**~upstream_master**

r00ta replied:
wooops! Thanks for the catch!

danielezonca commented on addons/monitoring-addon/monitoring-prometheus-addon/monitoring-prometheus-quarkus/src/main/resources/META-INF/kogito.addon:
```diff
@@ -0,0 +1 @@
+prometheus-monitoring
```
I'm not completely sure we need to decouple `monitoring` and `prometheus-monitoring` addons but it is probably fine

MarianMacik commented on addons/monitoring-addon/monitoring-prometheus-addon/monitoring-prometheus-quarkus/pom.xml:
```diff
@@ -0,0 +1,27 @@
+    <dependency>
+      <groupId>org.kie.kogito</groupId>
+      <artifactId>monitoring-core-quarkus</artifactId>
+    </dependency>
```
It would be good to put here quarkus-resteasy as well because we use it directly to define the `QuarkusMetricsResource`. Same thing is done with spring boot prometheus addon. It is always better to declare dependencies which are used and not rely on the transitive ones I think.

MarianMacik commented on addons/monitoring-addon/monitoring-core/monitoring-core-quarkus/src/main/java/org/kie/kogito/monitoring/core/quarkus/QuarkusEventListenerFactory.java:
```diff
@@ -0,0 +1,50 @@
+    @IfBuildProperty(name = Constants.MONITORING_PROCESS_USE_DEFAULT, stringValue = "true", enableIfMissing = true)
+    public DefaultProcessEventListenerConfig produceProcessListener() {
+        LOGGER.debug("Producing default listener for process monitoring.");
+        return new MonitoringProcessEventListenerConfig();
```
Just ony hypothetical question. If I add 2 addons with listeners (for example in future), will I have to solve this wiring myself? As then I need to somehow combine 2 listener configs into one, right?

r00ta replied:
The user can add any number of listeners without any limitation. We also give the possibility to disable this listener using the config property `Constants.MONITORING_PROCESS_USE_DEFAULT = false`, is this what you mean?

MarianMacik replied:
So in case I need to combine listeners from more than one addon, I set that property to false and add listeners manually, did I get it right?

danielezonca replied:
The idea is to let user disable default behavior and provide a different impl. All the listeners in general are wired using CDI and user can already add as many listeners as he wants (without the need to disable the default behavior).
Here the idea is just to opt-out default metric listeners

MarianMacik replied:
I see, so I can add listeners via CDI even if I go for `MonitoringProcessEventListenerConfig`, right?

danielezonca replied:
Yes, all listeners are collected and registered, this is just another one :)

MarianMacik reviewed (commented):
Looks good, just wrote down possible minor enhancements.

MarianMacik commented on addons/monitoring-addon/monitoring-core/monitoring-core-quarkus/pom.xml:
```diff
@@ -0,0 +1,39 @@
+    <dependency>
+      <groupId>org.kie.kogito</groupId>
+      <artifactId>kogito-api</artifactId>
+    </dependency>
```
I think kogito-api is also not needed here?

MarianMacik replied:
Sorry, my bad, actually the DefaultProcessEventListenerConfig is coming from the jbpm-flow, so that one should be there, but I would leave it as is now :)

r00ta commented:
Jenkins retest this please

r00ta commented:
@danielezonca @MarianMacik modules have been renamed according to https://github.com/kiegroup/kogito-examples/pull/458#discussion_r539376525
